### DISCUSSION TITLE
Meg's Default Template

### DISCUSSION THREAD

**Discussion by megclaypool** [SilentVoid13/Templater#259] [Templates Showcase]
I use this as my default template. If I created the page by clicking a previously unresolved link, it uses the link title as the page title. If it's just a new page, it'll prompt for the title. Then it fills in the page info with the title and creation date:
```
<%*
  let title = tp.file.title
  if (title.startsWith("Untitled")) {
    title = await tp.system.prompt("Title");
    await tp.file.rename(`${title}`);
  } 
  tR += "---"
%>
title:  <%* tR += `${title}` %>
created: <% tp.date.now("YYYY-MM-DD") %>
---
# <%* tR += `${title}` %>

<% tp.file.cursor() %>
```
('m outputting the triple dash with the templater return to avoid an extra line at the top of my page)

**Comment by MaksimZinovev** [comment]
Thank you! I could not figure out how to avoid an extra lines at the top of my script :)

**Reply by MaksimZinovev** [reply]
@megclaypool could you please explain, why do you use dollar sign and curly braces here `${title}` ?

**Reply by megclaypool** [reply]
It's a totally unnecessary [template literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) left over from a few edits ago. Since I ended up *only* outputting the title and no additional text, it would probably be better to just use `title:  <%* tR += title %>`, but it wasn't hurting anything so I didn't bother to change it 😆

(At one time I was outputting ``<%* tR += `title:  ${title}` %>`` instead 🤷🏽‍♀️)

**Reply by iwconfig** [reply]
`tR += "---"` isn't necessary. To remove one newline after the snippet, add a dash to the closing tag, like so: `-%>`. [Read more about whitespace control here.](https://silentvoid13.github.io/Templater/docs/commands/whitespace-control)

**Reply by iwconfig** [reply]
For example

    <%* title = 'blah' _%>
    
              
       
    ---
    <%* const foo = 'bar' %>
    title: <% title %>   FOO       
    
    
                                
    <%_ foo -%>
    
    ---

would become

    ---
    
    title: blah   FOObar
    ---

**Reply by megclaypool** [reply]
I just found that on another post -- thanks so much for mentioning it here! 😀

**Comment by megclaypool** [comment]
Here's a cleaned up version 😀 

```
<%*
  let title = tp.file.title
  if (title.startsWith("Untitled")) {
    title = await tp.system.prompt("Title");
    await tp.file.rename(title);
  } 
  tR += "---"
%>
title:  <%* tR += title %>
created: <% tp.date.now("YYYY-MM-DD") %>
---
# <%* tR += title %>

<% tp.file.cursor() %>
```

**Reply by midas0441** [reply]
I tried to extend your template with `if-then` statements between:
```
title = await tp.system.prompt("Title");
/* insert if-then */
await tp.file.rename(title);
```
The `if-then` defines values for `tags` and redefines `title` with a date-string --- all based on user input for `title`. Unfortunately, I'm having problems with the code running asynchronously. It seems to execute `await tp.file.rename(title)` without waiting for redefined `title` from the `if-then` statements. Googling shows a bunch of topics on callback, promise, await and so on.... all rather complicated to just get code to run sequentially. Is this expected javascript behavior?

**Reply by zabouti** [reply]
> <%*
>   let title = tp.file.title
>   if (title.startsWith("Untitled")) {
>     title = await tp.system.prompt("Title");
>     await tp.file.rename(title);
>   } 
>   tR += "---"

Meg, I'm an obsidian newbie with only rudimentary javascript skills.  I'm puzzled by the command where you append "---" to **tR**.  What I don't understand is where **tR** gets a value to append to?  It looks like an uninitialized variable to me, but I suspect that it is not.

Thanks for posting this.  I'm going to try using this code.

ge, chapel hill, nc

**Comment by megclaypool** [comment]
As I understand it, from our point of view tR already exists (so we don't have to declare it) -- it's the replacement string/return value created by the Eta template language used by Templater...

From [the documentation](https://silentvoid13.github.io/Templater/docs/commands/execution-command):
> Sometimes, you may want to output something when using a JS execution command. This is possible although a raw display / interpolation command is preferable.
> When the Eta templating engine generates a replacement string using all of our commands results, it is stored in a variable named `tR`. This is the string that will contain the processed file content. You are allowed to access that variable from a JS execution command.
> This means that, to output something from a JS execution command, you just need to append what you want to output to that tR string variable.
> For example, the following command: `<%* tR += "test" %>` will output `test`.

**Reply by zabouti** [reply]
Thanks, Meg,  This makes perfect sense but it also means I've got something else to learn.  ;-).

**Reply by iwconfig** [reply]
Using https://eta.js.org/playground you can actually see the generated function which might help with understanding how it works. Use regular Eta syntax and you should get the sense of it. I find that writing Templater stuff with the starting tag set to `<%*`, syntax which I believe is specific to Templater and not Eta, does not seem to work properly.

**Reply by megclaypool** [reply]
Ooo -- thanks! That's a cool resource! 😀

**Comment by patowod** [comment]
This is a great example to get started @megclaypool - I have one issue/question though and am wondering if anyone else has experienced this. I get two prompts for the title. The input of the second prompt is then used. I can't wrap my head around it. Any help would be appreciated.

**Comment by namtrah** [comment]
How did you actually implement this (and yes, I am a noob to Obsidian). Sorry if this seems obvious, but it is not for me.

**Reply by seefrye** [reply]
1. Install Templater in the Obsidian plug-in finder. I'm assuming you got this far, but if not just open Settings > Community Plugins and search for "Templater", and then Enable it once it installs.
2. Still in Settings, go to the settings for Templater. Set a directory where you will keep your templates.
3. Copy/Paste the code above (make sure you look in the comments for an updated version) into a new blank file in your Templates directory. Delete the leading and trailing backticks.
4. Create another new note, then click the Templater button on the left-hand menu - it looks like this ---> \<\%
5. This will pull the template, and it will prompt you for a title.

**Comment by gmcusaro** [comment]
Is there way to check empty prompt and replace with something by default?
Like: ```title = await tp.system.prompt("Title") ?? "YOURDEFAULTVALUE";```

With ``` ?? "String"``` I get this error:
<img width="320" alt="error" src="https://user-images.githubusercontent.com/30832514/139603681-24e1ac9a-36fc-4e1b-96d5-d95efe1344a8.png">

**Reply by iwconfig** [reply]
replace `??` with `||`, or it's opposite `&&`.

**Reply by iwconfig** [reply]
~~`??` isn't valid JavaScript.~~ (see edit below)

Maybe you're thinking of the shorthand for if-then statements which is `(condition) ? true : false`

Using this in your case would have to be something like

```javascript
let title = await tp.system.prompt("Title")
title = title ? title : 'YOURDEFAULTVALUE'
```

which is the same as 
```javascript
let title = await tp.system.prompt("Title")
if (title) {
	title = title
} else {
	title = 'YOURDEFAULTVALUE'
}
```
But that's not particularly elegant, in this case, compared to just using logical operators.

- `&&` meaning logical AND
- `||` meaning logical OR

---
EDIT: The [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) (`??`) only returns its right-side operand if the left-side is either `null` or `undefined`. I could be wrong but I suspect `tp.system.prompt` doesn't return either of these, and thus it fails due to the [operator precedence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence).

> The nullish coalescing operator has the fifth-lowest operator precedence, directly lower than || and directly higher than the conditional (ternary) operator.

**Reply by iwconfig** [reply]
This works because `null`, `undefined`, `NaN`, `0` and `''` (empty string) also equates to `false`.

**Reply by gmcusaro** [reply]
Great, thanks!

**Reply by iwconfig** [reply]
Actually, I just did a proper test myself. Turns out I was wrong all along. The nullish coalescing operator works fine. Sorry! You just have to place the code inside a [execution tag](https://silentvoid13.github.io/Templater/docs/commands/execution-command) (`<%*`). I suspect you tried executing it inside a normal [raw display tag](https://silentvoid13.github.io/Templater/docs/commands) (`<%`).

So to sum it up, just do this:
```js
<%*
title = await tp.system.prompt("Title") ?? "YOURDEFAULTVALUE"
// do whatever
%>
```

(notice the opening tag)

**Comment by mjh8431** [comment]
Thanks for this template. I have some questions about it. I am new to this so I hope you can help me.
First, when I copied and pasted it in to a template note this part got formatted as a header:  
**created: <% tp.date.now("YYYY-MM-DD") %>**
But I don't see any way to make it normal text.  It is normal text in the output. Any ideas? Maybe it doesn't matter but it looks strange in the template.
My second question: It there a way to add the date to the Title, as in "YYYYDDMMHHss" "Title" ?
Third, I have some other items in my YAML frontmatter.
    tags:
    type:
    quote:
    paraphrase:
How an I set up the template to allow me to fill in the information like is done to change the title? Again, I am new to this so I don't quite understand how to do it.

Thanks

**Comment by Adito5393** [comment]
Thanks for sharing the template! Here's my addition for the same template when I forget to generate it for the dat/week/time it was meant to be generated:

```templater
<%*
// THIS TEMPLATER will change the Time Source so that it generates a file from the Future / Past!

let originalToday = tp.date.now("YYYY,MM,DD");
let refDate = tp.date.now("YYYY,MM,DD");

const choices = [tp.date.now("YYYY,MM,DD",+1), tp.date.now("YYYY,MM,DD",+2), "placeholder"];
refDate = await tp.system.suggester(choices, choices);
if (refDate == "placeholder") {
	refDate = await tp.system.prompt("Reference date of the week to generate (format: YYYY,MM,DD)");
}
moment.now = function () {
	return +new Date(refDate);
}

// Remember to reset the day at the end of file!
// REST of code from day templateR
...

<%*
// DO NOT DELETE (the week TEMPLATER does not need this)
//  Restore today day to moment JS !!!
moment.now = function () {  return +new Date(originalToday);  }
-%>
```

This will offer a dropdown menu to choose the date in the correct format or enter it manually:
![image](https://github.com/SilentVoid13/Templater/assets/39572144/ee690463-d2c0-4e9e-bdc3-7e67a34a4aa9)

**Reply by megclaypool** [reply]
Nice! I use the calendar plugin to generate daily notes with dated file titles that I can use to get the appropriate date, but this is a cool alternative 😎

**Reply by Adito5393** [reply]
The calendar plugin works great, indeed, for the daily notes! But it doesn't work for weekly overview files... so I use this approach for the ultimete flexibility 😄

**Comment by guptarohit** [comment]
Thanks for this Meg 😃

I found that cursor jump is not working anymore i.e. `<% tp.file.cursor() %>` https://github.com/SilentVoid13/Templater/issues/397

Using following instead for rescue:
`<%*app.workspace.activeLeaf.view.editor?.focus(); %>`

My default template:
```
<%*
  let title = tp.file.title
  if (title.startsWith("Untitled")) {
    title = await tp.system.prompt("Title");
    await tp.file.rename(title);
  } 
  tR += "---"
%>
title:  <%* tR += title %>
created: <% tp.date.now("YYYY-MM-DD HH:mm") %>
updated: <% tp.file.last_modified_date("YYYY-MM-DD HH:mm") %>
tags: 

---
# <%* tR += title %>
<%* app.workspace.activeLeaf.view.editor?.focus(); %>
```

**Reply by Gryn23** [reply]
It still works, as stated in the text on your link :) You need to re-enable cursor jump though in Templater settings, as it is turned off by default now.

**Comment by galinT** [comment]
This is how I do it using [Whitespace Control](https://silentvoid13.github.io/Templater/commands/whitespace-control.html#whitespace-control) and [Script User Functions](https://silentvoid13.github.io/Templater/user-functions/script-user-functions.html#script-user-functions)

```js
async function promptTitle(tp, journalType){
    let title = tp.file.title;
    if (title.startsWith("Untitled")) {
        title = await tp.system.prompt("Title?");
    }

    if(journalType){
        title = `${tp.file.creation_date("YYYY-MM-DD")} - ${journalType} - ${title}`;
    }
   
    await tp.file.rename(title);
    return title;
}

module.exports = promptTitle;
```

```yaml
<%* let title = await  tp.user.promptTitle(tp, "Meeting"); -%>
---
title:  <% title %>
tags:
  - log/meeting
---
```
